output/cloud/expv2/metrics_client.go

```diff
@@ -0,0 +1,127 @@
+package expv2
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/snappy"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/lib/consts"
+	"go.k6.io/k6/output/cloud/expv2/pbcloud"
+)
+
+type httpDoer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// MetricsClient is a Protobuf over HTTP client for sending
+// the collected metrics from the Cloud output
+// to the remote service.
+type MetricsClient struct {
+	httpClient httpDoer
+	logger     logrus.FieldLogger
+	token      string
+	userAgent  string
+
+	pushBufferPool sync.Pool
+	baseURL        string
+}
+
+// NewMetricsClient creates and initializes a new MetricsClient.
+func NewMetricsClient(logger logrus.FieldLogger, host string, token string) (*MetricsClient, error) {
+	if host == "" {
+		return nil, errors.New("host is required")
+	}
+	if token == "" {
+		return nil, errors.New("token is required")
+	}
+	return &MetricsClient{
+		httpClient: &http.Client{Timeout: 5 * time.Second},
+		logger:     logger,
+		baseURL:    host + "/v2/metrics/",
+		token:      token,
+		userAgent:  "k6cloud/v" + consts.Version,
+		pushBufferPool: sync.Pool{
+			New: func() interface{} {
+				return &bytes.Buffer{}
+			},
+		},
+	}, nil
+}
+
+// Push pushes the provided metrics the given test run.
+func (mc *MetricsClient) Push(ctx context.Context, referenceID string, samples *pbcloud.MetricSet) error {
+	if referenceID == "" {
+		return errors.New("TestRunID of the test is required")
+	}
+	start := time.Now()
+
+	b, err := newRequestBody(samples)
+	if err != nil {
+		return err
+	}
+
+	buf, _ := mc.pushBufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer mc.pushBufferPool.Put(buf)
+
+	_, err = buf.Write(b)
+	if err != nil {
+		return err
+	}
+	// TODO: it is always the same
+	// we don't expect to share this client across different refID
+	// with a bit of effort we can find a way to just allocate once
+	url := mc.baseURL + referenceID
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, buf)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("User-Agent", mc.userAgent)
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("Content-Encoding", "snappy")
+	req.Header.Set("K6-Metrics-Protocol-Version", "2.0")
+	req.Header.Set("Authorization", "Token "+mc.token)
+
+	resp, err := mc.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if err := cloudapi.CheckResponse(resp); err != nil {
+		return err
+	}
+	mc.logger.WithField("t", time.Since(start)).WithField("size", len(b)).
+		Debug("Pushed the collected metrics to the Cloud service")
+	return nil
+}
+
+const payloadSizeLimit = 100 * 1024 // 100 KiB
```

Comment:
Shouldn't we just break them in parts then? 100kb doesn't soudn *that* big. Especially before compression.

Reply:
How would you do that? Here the limit is related to the protobuf, how do you chunk it without breaking the metric set message? I'm not aware of Protobuf supporting a streaming/chunked version. Did I miss it?

This should be an extreme check, if we need to chunk, then I think we should do that before calling the metric's client, so in its caller. WDYT?

Reply:
You can use https://pkg.go.dev/google.golang.org/protobuf/proto#Size to find out how big it will be *before* serializing and then splitting if necessary.

And I guess a method like Push can know that protobuf will be used and cut it in pieces when necessary.

We can probably leave that for after we merge everything. But I would like to know how many metrics are 100kb before we actually make this the default. As dropping your metrics because they were 101kb sounds like terrible experience.